firestore/src/jni/arena_ref.cc

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <atomic>
+
+#include "firestore/src/jni/arena_ref.h"
+
+#include "firestore/src/jni/hash_map.h"
+#include "firestore/src/jni/long.h"
+#include "firestore/src/jni/map.h"
+
+namespace firebase {
+namespace firestore {
+namespace jni {
+
+namespace {
+
+HashMap* gArenaRefHashMap = nullptr;
+
+int64_t GetNextArenaRefKey() {
+  static std::atomic<int64_t> next_key(1);
+  return next_key.fetch_add(1);
+}
+
+}  // namespace
+
+ArenaRef::ArenaRef(Env& env, const Object& object)
+    : key_(GetNextArenaRefKey()) {
+  gArenaRefHashMap->Put(env, key_object(env), object);
+}
+
+ArenaRef::ArenaRef(const ArenaRef& other) : key_(GetNextArenaRefKey()) {
+  if (other.key_ != 0) {
+    Env env;
+    Local<Object> object = other.get(env);
+    gArenaRefHashMap->Put(env, key_object(env), object);
+  }
+}
+
+ArenaRef::ArenaRef(ArenaRef&& other) {
+  key_ = other.key_;
+  other.key_ = 0;
+}
+
+ArenaRef& ArenaRef::operator=(const ArenaRef& other) {
+  Env env;
+
+  if (key_ != other.key_) {
```

Comment:
Rather than checking if `if (key_ != other.key_)`, check if `this == &other`. This is more conventional in copy assignment operator overloads. In fact, I'd put that as the first thing in this method, to reduce nesting and improve readability.

i.e.

```
ArenaRef& ArenaRef::operator=(const ArenaRef& other) {
  if (this == &other) {
    return *this;
  }
  ...
```

Please do the same thing for the move assignment operator.